// @mintplex-labs/node-mic-monitor/index.js
const { platform, arch } = require('os');
const EventEmitter = require('events');
const path = require('path');
const fs = require('fs');

/**
 * @typedef {Object} MicrophoneMonitorOptions
 * @property {number} [timeout=5_000] - The timeout in milliseconds to wait before checking if the microphone is active. Defaults to 5 seconds.
 * @property {Array<MicrophoneMonitorEventsProcessDefinition>} [processFilters=[]] - The process filters to apply to the microphone activity to be ignored as valid microphone activity. Defaults to an empty array.
 * @property {boolean} [matchOnlyFilters=false] - Change behaviour use `processFilters` to only give back processes that match the filters instead of ignoring processes that match the filters. Defaults to false.
 * @property {string|null} [_binFolderOverride=null] - The path to the binary folder to override the default binary folder. Defaults to null.
 * @property {boolean} [_debug=false] - Whether to enable debug mode. Defaults to false.
 * @property {boolean} [_debugInterval=false] - Whether to enable debug mode for the interval. Defaults to false. Must be used in conjunction with _debug.
 */

/**
 * @typedef {Object} AudioProcessUtilResponse
 * @property {string} processName - The name of the process.
 * @property {number} processPID - The PID of the process.
 */

/**
 * @typedef {Object} FormattedAudioProcess
 * @property {string} name - The name of the process (always lowercase).
 * @property {number} pid - The PID of the process.
 */

/**
 * @typedef {Object} MicrophoneActiveEvent
 * @property {Array<FormattedAudioProcess>} audioProcesses - The names of the audio processes.
 */

/**
 * @typedef {Object} MicrophoneMonitorEventsProcessDefinition
 * Will be used on process detection to pre-filter the processes you wish to **ignore** as valid microphone activity
 * @property {'static'|'regex'} type - The type of the process definition.
 * @property {string | RegExp} value - The value of the process definition.
 */

/**
 * Map of known popular process names to their human-readable names
 * Contains both mac and windows process names
 * for windows, this is just the name of the executable file
 * @type {Record<string, string>}
 */
const KNOWN_MAPS = {
  // Chrome
  'com.google.chrome': 'Google Chrome',
  'com.google.chrome.helper': 'Google Chrome',
  'chrome': 'Google Chrome',

  // Safari
  'com.apple.safari': 'Safari',
  'com.apple.webkit': 'Safari',
  'com.apple.webkit.gpu': 'Safari',

  // Voice Memo/Recorder
  'com.apple.voicemos': 'Voice Memos',
  'voicerecorder': 'Voice Recorder',

  // Microsoft Teams
  'com.microsoft.teams2': 'Microsoft Teams',
  'com.microsoft.teams': 'Microsoft Teams',
  'ms-teams': 'Microsoft Teams',

  // Zoom
  'us.zoom.xos': 'Zoom',
  'zoom': 'Zoom',

  // Slack
  'com.tinyspeck.slackmacgap.helper': 'Slack',
  'com.tinyspeck.slackmacgap': 'Slack',
  'slack': 'Slack',

  // Discord
  'com.hnc.discord.helper.renderer': 'Discord',
  'com.hnc.discord.helper': 'Discord',
  'com.hnc.discord': 'Discord',
  'discord': 'Discord',
}

class MicrophoneMonitor {
  static Events = {
    /**
     * The event emitted when the microphone is active.
     */
    MICROPHONE_ACTIVITY_DETECTED: 'microphone-activity-detected',
    /**
     * The event emitted when the microphone is stopped.
     */
    MICROPHONE_ACTIVITY_STOPPED: 'microphone-activity-stopped'
  };

  emitter;
  timeout;
  interval;
  platform_utils;
  isMicrophoneActive;
  filters;

  /**
  * Creates a new MicrophoneMonitor instance
  * @param {MicrophoneMonitorOptions} options - The options for the MicrophoneMonitor instance
  */
  constructor({
    timeout = 5_000,
    processFilters = [],
    matchOnlyFilters = false,
    _binFolderOverride = null,
    _debug = false,
    _debugInterval = false
  } = {}) {
    this.timeout = timeout;
    this.interval = null;
    this.emitter = new EventEmitter();
    this.isMicrophoneActive = false;
    this.filters = processFilters;
    this._binFolderOverride = _binFolderOverride;
    this._debug = _debug;
    this.watchingPid = null;
    this.debugInterval = _debugInterval;
    this.matchOnlyFilters = matchOnlyFilters;

    if (!['darwin', 'win32'].includes(process.platform)) throw new Error('Microphone monitoring is only supported on macOS and Windows');
    this.platform_utils = this.#tryRequirePlatformUtils();
    this.#log('initialized', JSON.stringify({ timeout, processFilters, matchOnlyFilters, _binFolderOverride, _debug }, null, 2));
  }

  #log(message, ...args) {
    if (this._debug) console.log('[MicrophoneMonitor]', message, ...args);
  }

  #tryRequirePlatformUtils() {
    const targetAddon = `${platform}_${arch}_utils.node`;
    try {
      if (this._binFolderOverride) return require(path.join(this._binFolderOverride, targetAddon));
    } catch (error) {
      throw error;
    }

    try {
      return require("bindings")(targetAddon);
    } catch (error) {
      // Try to require the addon from root directory
      let addonPath = path.join(__dirname, targetAddon)
      // If the addon is not found in the root directory, try to require it from the lib directory
      if (!fs.existsSync(addonPath)) addonPath = path.join(__dirname, 'lib', targetAddon)
      // If the addon is not found in the lib directory, throw an error
      if (!fs.existsSync(addonPath)) throw new Error(`Addon ${targetAddon} not found`);
      return require(addonPath);
    }
  }

  /**
   * Filters the process name from its path eg: /Applications/Google Chrome.app/Contents/MacOS/Google Chrome -> google chrome
   * @param {AudioProcessUtilResponse} inputAudioProcessesPath - The path to the input audio process
   * @returns {FormattedAudioProcess}
   */
  _formatAudioProcesses(audioProcessUtilResponse) {
    const { processName, processPID = null } = audioProcessUtilResponse;
    if (!processName) return { name: null, pid: null };

    // Only on Windows, the process name is the full path to the process
    // eg: C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe -> chrome
    if (process.platform === 'win32') {
      return {
        name: path.win32.basename(processName).replace(path.extname(processName), '').toLowerCase(),
        pid: isNaN(processPID) ? null : processPID
      }
    }

    // On macOS, the process name is the program name
    // eg: com.apple.VoiceMemos 
    if (process.platform === 'darwin') {
      return {
        name: processName,
        pid: isNaN(processPID) ? null : processPID
      }
    }
    return { name: null, pid: null }
  }

  /**
   * Applies the process filters to the input audio process util response
   * If `matchOnlyFilters` is true, only returns processes that match the filters
   * If `matchOnlyFilters` is false, returns all processes that do not match the filters (default behaviour)
   * @see {MicrophoneMonitorEventsProcessDefinition}
   * @param {Array<FormattedAudioProcess>} processes - The input processed audio process
   * @returns {Array<FormattedAudioProcess>}
   */
  _applyProcessFilters(_processes = []) {
    // Filter out processes that dont have a name
    const readableProcesses = _processes.length ? _processes.filter((p) => typeof p.name === 'string') : _processes;
    if (!this.filters.length) return readableProcesses;
    const unfilteredProcesses = [];
    const matchedProcesses = [];

    for (const micProcess of readableProcesses) {
      const staticFilter = this.filters.find(filter => filter.type === 'static' && filter.value === micProcess.name);
      if (staticFilter) {
        if (!this.matchOnlyFilters) this.#log('Filtering out process due to static filter match', micProcess.name, staticFilter.value);
        else this.#log('Adding process to matched processes due to static filter match', micProcess.name, staticFilter.value);
        matchedProcesses.push(micProcess);
        continue;
      }

      const regexFilter = this.filters.find(filter => filter.type === 'regex' && filter.value.test(micProcess.name));
      if (regexFilter) {
        if (!this.matchOnlyFilters) this.#log('Filtering out process due to regex filter', micProcess.name, regexFilter.value);
        else this.#log('Adding process to matched processes due to regex filter', micProcess.name, regexFilter.value);
        matchedProcesses.push(micProcess);
        continue;
      }

      unfilteredProcesses.push(micProcess);
    }

    return this.matchOnlyFilters ? matchedProcesses : unfilteredProcesses;
  }

  /**
  * Starts watching for microphone activity for a SINGLE PROCESS ONLY specific PID (should have come from the microphone-activity-detected event)
  * This will be cleaned up automatically when the microphone is stopped or the specific PID we target is no longer running
  * @param {number} pid - The PID to watch for microphone activity
  */
  watchForSpecificPid(pid) {
    this.#log('Watching for microphone activity for specific PID', pid);
    this.watchingPid = pid;
  }

  /**
   * Starts watching for microphone activity
   */
  watchForMicrophoneActive() {
    this.#log('Watching for microphone activity...');
    this.interval = setInterval(() => {
      /** @type {Array<AudioProcessUtilResponse>} */
      const inputAudioProcesses = this.platform_utils.getRunningInputAudioProcesses()
      const formattedProcesses = this._applyProcessFilters(inputAudioProcesses.map((p) => this._formatAudioProcesses(p)));
      if (this.debugInterval) this.#log('Available input audio processes:', JSON.stringify({
        all: inputAudioProcesses,
        filtered: formattedProcesses
      }, null, 2));

      // If the microphone is not active and we have processes - emit the microphone-newly-active event as microphone is newly active
      if (!this.isMicrophoneActive) {
        if (formattedProcesses.length === 0) return;
        this.emitter.emit(MicrophoneMonitor.Events.MICROPHONE_ACTIVITY_DETECTED, formattedProcesses);
        this.isMicrophoneActive = true;
        return;
      }

      // If we are watching for a specific PID and we can no longer find it in the available processes, stop watching for microphone active
      if (
        this.watchingPid &&
        !formattedProcesses.find(p => p.pid === this.watchingPid)
      ) return this.stopWatchingForMicrophoneActive();

      // If the processes are empty, stop watching for microphone active 
      if (formattedProcesses.length === 0) return this.stopWatchingForMicrophoneActive();
    }, this.timeout);
  }

  /**
  * Stops watching for microphone activity
  * @param {boolean} [restart=false] - Whether to restart watching for microphone activity. Defaults to false.
  */
  stopWatchingForMicrophoneActive(restart = false) {
    this.#log('Cleaning up microphone monitoring...');
    if (this.interval) clearInterval(this.interval);
    this.interval = null;
    if (!this.isMicrophoneActive) return;
    this.emitter.emit(MicrophoneMonitor.Events.MICROPHONE_ACTIVITY_STOPPED);
    this.emitter.removeAllListeners();
    this.emitter = null;
    this.isMicrophoneActive = false;
    this.watchingPid = null;
    this.#log('Microphone monitoring cleaned up');
    if (restart) this.watchForMicrophoneActive();
  }

  /**
   * Humanizes the process name from the internal list of known process names
   * @param {string} processName - The process name to humanize
   * @returns {string}
   */
  static humanizeProcessName(processName) {
    return KNOWN_MAPS[processName.toLowerCase()] ?? processName;
  }
}

module.exports = MicrophoneMonitor;
